### ISSUE TITLE
[RFC]: Experimental Recirculation for causal decoders

### ISSUE THREAD

**Issue by mihir-s-05** [vllm-project/vllm#53401]
### Motivation.

[Recirculation](https://arxiv.org/abs/2608.17981) is an inference-time method
that changes a pretrained causal decoder's effective depth without changing
its weights. A normal pass produces logits and a deep residual-stream state.
That state is norm-matched, mixed into an earlier residual boundary, and the
upper stack is run again. The rerun overwrites upper-layer KV entries so later
tokens attend to the recirculated representation.

The design question for vLLM is broader than adding a branch to one model's
`forward`: cache ownership, scheduler behavior, CUDA-graph shapes, model
capability negotiation, and recurrent state must agree on the same semantics.
This RFC asks whether vLLM wants that behavior as an experimental engine
feature with explicit model-family opt-in.

This proposal supersedes author-closed RFC #53178. That issue was closed before
feedback and received no maintainer comments; it should not be interpreted as
either approval or rejection. This proposal narrows the initial code review to
the core engine and Gemma 3, with other model families reviewed separately.

### Algorithm and cache invariant

For destination layer `d` and source layer `s`, where
`0 <= d < s < num_hidden_layers`, one scheduled token block executes as:

```text
normal pass: layers 0 ... d ... s ... N
                  |       |
                  |       +-- source residual
                  +---------- destination residual

mix: norm_match(source, destination) * alpha + destination * beta

rerun: mixed residual -> layers d+1 ... N
```

The normal pass's final hidden state supplies logits. The rerun replaces the
block's upper-layer cache state; lower-layer cache entries are unchanged. A
one-token block implements tokenwise Recirculation. Larger prefill blocks are
an explicit blockwise approximation: their logits come from the normal pass,
and their recirculated cache affects later blocks.

### Experimental configuration and capability boundary

The initial implementation accepts a validated dictionary through
`--hf-overrides`, avoiding a stable public CLI contract while the feature is
experimental:

```json
{
  "recirculation_config": {
    "source_layer": 11,
    "destination_layer": 4,
    "alpha": 0.15,
    "beta": null,
    "ramp_tokens": 10,
    "wavefront": true
  }
}
```

Validation covers unknown or missing fields, layer ordering, coefficient
ranges, identity configurations, ramping, and both ordinary and MRoPE token
positions. An open question is whether maintainers prefer a first-class
configuration object before merge.

Engine code detects an explicit `SupportsRecirculation` capability. Models opt
in through a shared residual-decoder mixin and declare the modes they support;
unreviewed subclasses and wrappers fail closed. Families may override layer
execution, residual materialization, finalization, or non-KV state handling
without duplicating the scheduler and cache protocol.

### Execution paths

The serial compatibility path runs the normal stack, records the source and
destination boundaries, returns the normal final state for logits, and reruns
the upper stack from the mixed residual. Ordinary attention KV entries are
overwritten at the same token positions. Hybrid recurrent models must instead
snapshot and restore affected convolutional or recurrent state before the
rerun.

An optional two-token wavefront batches the previous token's recirculated
upper pass with the current token's normal upper pass after a one-token warmup:

```text
row 0: previous token's recirculated upper pass
row 1: current token's normal upper pass
```

At every upper attention layer, row 0 first replaces the previous token's KV
entry and row 1 then consumes the same effective cache as serial execution.
The engine owns the pending residual, positions, attention metadata, remapped
slot mappings, and a dedicated two-row upper-stack CUDA graph.

The first wavefront implementation is limited to one sequence, one scheduled
token, FlashAttention, and no prefix caching or distributed parallel modes.
Other compatible configurations use the serial path.

### Safety boundary

The initial implementation targets Model Runner V1 and rejects explicit V2
selection. It also rejects speculative decoding and auxiliary hidden-state
extraction, pipeline parallelism, non-causal or diffusion backbones,
DeepSeek-V4 hyperconnection state, unreviewed wrappers, and parallel modes
whose state ownership has not been defined. Speculative decoding would need a
separate definition of which pass owns draft blocks, cache writes, and
auxiliary states.

MoE routing is recomputed during the upper rerun. A mixed residual may select
different experts; the proposal does not replay fixed routing decisions.

### Staged rollout

Implementation is split into dependent, scoped PRs:

1. Core engine/configuration and Gemma 3.
2. Ordinary residual dense and MoE adapters.
3. DeepSeek-V3.2 and GLM DSA adapters.
4. Qwen3-Next and Qwen 3.5 recurrent-state adapters.
5. MiniMax-M3 sparse-attention adapters.
6. Kimi Linear and reviewed wrapper delegation.

Each family opts in only after its residual, cache, and non-KV state semantics
are reviewed and tested; support is not inferred from model configuration.
Kimi-K3 AttnRes is excluded because its residual-bank mapping requires a
separate design decision.

Serial Recirculation intentionally adds upper-stack compute and cache writes.
Wavefront execution reduces dispatch overhead, not mathematical work. No
universal quality or performance claim is proposed: each adapter must provide
correctness tests and appropriate real-model evidence.

### Feedback Period.

Fourteen calendar days, with no code PR until a maintainer indicates that the
core direction is worth reviewing. Main questions:

1. Is explicit capability negotiation plus a shared residual-decoder mixin the
   right boundary?
2. Should Llama and Mistral accompany Gemma 3 in the core PR or follow it?
3. Is V1-only experimental support acceptable, or should this wait for V2?
4. Are the wavefront invariant and initial restrictions convincing?
5. Should configuration use `hf_overrides` or a first-class interface?
6. What real-model evidence should be required for later adapter classes?

### Any Other Things.

- Paper: <https://arxiv.org/abs/2608.17981>
- Superseded author-closed RFC: #53178
- Superseded author-closed draft PR: #53179

The local prototype has exercised baseline, identity, serial, and wavefront
generation on an RTX 3080 using dummy weights through eager and compiled engine
paths, FlashAttention 2, CUDA graphs, and slot mapping. A cached Gemma 3 1B
sanity check over 1,020 scored tokens measured compiled perplexity `36.9618`
for baseline, `32.5482` for serial, and `32.2460` for wavefront. This was a
small fixed sample, not a standardized benchmark, and must be rerun on the
final rebased PR head.

Core acceptance requires parser and mixing tests, adapter cache/state tests,
serial/wavefront equivalence tests, direct runner and slot-mapping coverage,
and at least one real-weight comparison on the final PR head. Large Kimi,
MiniMax-M3, GLM-5, and similar checkpoints have not received local real-weight
validation; synthetic tests are not a substitute, so those follow-up PRs need
fresh model evidence or maintainer-approved provider validation.

### AI assistance disclosure

OpenAI Codex provided non-trivial assistance with implementation, code review,
test development, local validation, branch decomposition, and this RFC draft.
The human submitter remains responsible for reviewing, understanding, and
defending every proposed change and reported result.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by mihir-s-05** [comment]
The 14-day feedback period has elapsed without comments. This is an engine-level proposal rather than a Kimi-specific request; the kimi label appears to have been added automatically.  
Before I rebase the implementation or open a PR, could an engine owner advise whether a serial-only core + Gemma 3 first PR is worth pursuing? I am happy to defer wavefront execution and all later model-family adapters to keep the initial review narrow.  
@njhill, tagging you because the initial design primarily affects the V1 worker/model-runner path.
